exotations/solvers/exotica_levenberg_marquardt_solver/src/LevenbergMarquardt.cpp

```diff
@@ -0,0 +1,138 @@
+
+    double lambda = 0;  // damping factor
+
+    int iterations_ = -1;
+};
+
+REGISTER_MOTIONSOLVER_TYPE("LMSolver", exotica::LevenbergMarquardt)
+
+void LevenbergMarquardt::Instantiate(LMSolverInitializer& init) { parameters_ = init; }
+void LevenbergMarquardt::specifyProblem(PlanningProblem_ptr pointer)
+{
+    if (pointer->type() != "exotica::UnconstrainedEndPoseProblem")
+    {
+        throw_named("This LevenbergMarquardt can't solve problem of type '" << pointer->type() << "'!");
+    }
+
+    MotionSolver::specifyProblem(pointer);
+
+    // generic problem
+    problem_ = pointer;
+
+    // specific problem
+    prob_ = std::static_pointer_cast<UnconstrainedEndPoseProblem>(pointer);
+}
+
+void LevenbergMarquardt::Solve(Eigen::MatrixXd& solution)
+{
+    prob_->resetCostEvolution(getNumberOfMaxIterations() + 1);
+
+    Timer timer;
+
+    if (!prob_) throw_named("Solver has not been initialized!");
+
+    const Eigen::VectorXd q0 = prob_->applyStartState();
+
+    if (prob_->N != q0.rows()) throw_named("Wrong size q0 size=" << q0.rows() << ", required size=" << prob_->N);
+
+    solution.resize(1, prob_->N);
+
+    lambda = parameters_.Damping;  // initial damping
+
+    const Eigen::MatrixXd I = Eigen::MatrixXd::Identity(prob_->Cost.J.cols(), prob_->Cost.J.cols());
+    Eigen::MatrixXd J;
+
+    Eigen::VectorXd q = q0;
+    double error = std::numeric_limits<double>::infinity();
+    double error_prev = std::numeric_limits<double>::infinity();
+    Eigen::VectorXd yd;
+    Eigen::VectorXd qd;
+    for (size_t i = 0; i < getNumberOfMaxIterations(); iterations_ = ++i)
+    {
+        prob_->Update(q);
+
+        yd = prob_->Cost.S * prob_->Cost.ydiff;
+
+        // weighted sum of squares
+        error_prev = error;
+        error = prob_->getScalarCost();
+
+        prob_->setCostEvolution(i, error);
+
+        J = prob_->Cost.S * prob_->Cost.J;
+
+        // source: https://uk.mathworks.com/help/optim/ug/least-squares-model-fitting-algorithms.html, eq. 13
+
+        if (i > 0)
+        {
+            if (error < error_prev)
+            {
+                // success, error decreased: decrease damping
+                lambda = lambda / 10.0;
+            }
+            else
+            {
+                // failure, error increased: increase damping
+                lambda = lambda * 10.0;
+            }
+        }
+
+        if (debug_) HIGHLIGHT_NAMED("Levenberg-Marquardt", "damping: " << lambda);
+
+        Eigen::MatrixXd M;
+        if (parameters_.ScaleProblem == "none")
+        {
+            M = I;
+        }
+        else if (parameters_.ScaleProblem == "Jacobian")
+        {
+            M = (J.transpose() * J).diagonal().asDiagonal();
+        }
+        else
+        {
+            throw std::runtime_error("no ScaleProblem of type " + parameters_.ScaleProblem);
+        }
+
+        qd = (J.transpose() * J + lambda * M).completeOrthogonalDecomposition().solve(J.transpose() * yd);
```

Comment:
`completeOrthogonalDecomposition` is sadly not available in 16.04's Eigen by default :'-(.

Reply:
That's a problem :-)

I used the LDLT decomposition before but got unstable results, i.e. different solutions (x) to the same problem (A,b). I haven't tried the LLT, but the instability with LDLT was really nasty and difficult to track down. I used sparse Jacobians back then and LDLT or LLT might work for other cases with less sparse Jacobians.
Because this issue is latent, I would generally recommend using the COD solver.

Options:
1. Do not enable/build on Ubuntu 16.04. Another good reason to upgrade to 18.04 :-)
2. Provide a Eigen catkin package with a sufficient high version. I don't really like this, but if someone really want to use this solver n 16.04.
3. Use a different library for solving. But this might result in similar issues as discussed with alternative decompositions.

Reply:
This is a tricky one:
(1) Easy to do with CMake, bad for user experience (people will ask 'why does this not work/is not available')
(2) May give us linking issues or runtime issues on release? Can we have it _exclusively_ as a private external for this project using `ExternalProject_Add` which is privately linked/not exposed?

Reply:
How do you add a private external dependency? CMake will fail for everyone that does not have access to the private repo.

I am in favour of option 1: Check the Eigen version and give a warning that this solver is not available for the poor souls that still use Ubuntu 16.04.

Reply:
I meant add the `ExternalProject_Add` to this project only. Do not use the `find_package(Eigen3)`. I am not 100% clear how to use the INTERFACE etc. features to avoid exposing Eigen-version-mismatch issues to other parts of the code, I am afraid.

> Check the Eigen version and give a warning that this solver is not available for the poor souls that still use Ubuntu 16.04.

Which, as of now, is every robot we have or are in the process of procuring. Upgrades to 18.04 can easily be 1-2 years out.

Reply:
I made the solver only built when Eigen version >= 3.3 is available. If someone wants to use this solver, he needs to manually install Eigen 3.3 or higher.

Reply:
Ubuntu 16.04 is using Eigen version 3.3 **beta1** (which has a serious bug: https://bugs.launchpad.net/ubuntu/+source/eigen3/+bug/1750794 you might want to be aware of).

The `CompleteOrthogonalDecomposition` was introduced in version 3.3 **beta2** with header `CompleteOrthogonalDecomposition.h`.
A fourth option would be to provide this header for users of version 3.3 beta1 (version 3.2.92 according to pkg-config and cmake).